broker.go

```diff
@@ -0,0 +1,710 @@
+package kafka
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/optiopay/kafka/proto"
+)
+
+const (
+	// StartOffsetNewest configures consumer to fetch messages produced after
+	// creating the consumer.
+	StartOffsetNewest = -1
+
+	// StartOffsetOldest configures consumer to fetch starting from the oldest
+	// message available
+	StartOffsetOldest = -2
+)
+
+var (
+	// Returned by consumer Fetch when retry limit was set and exceeded during
+	// single function call
+	ErrNoData = errors.New("no data")
+)
+
+type Logger interface {
+	Print(...interface{})
+	Printf(string, ...interface{})
+}
+
+type clusterMetadata struct {
+	created   time.Time
+	nodes     map[int32]string // node ID to address
+	endpoints map[string]int32 // topic:partition to leader node ID
+}
+
+type BrokerConf struct {
+	// Kafka client ID.
+	ClientID string
+
+	// LeaderRetryLimit limits number of connection attempts to single node
+	// before failing. Use LeaderRetryWait to control wait time between
+	// retries. Defaults to 10.
+	LeaderRetryLimit int
+
+	// LeaderRetryWait controls duration of wait between trying to connect to
+	// single node after failure. Defaults to 500ms.
+	// Timeout on connection is controlled by DialTimeout setting.
+	LeaderRetryWait time.Duration
+
+	// Any new connection dial timeout. By default 10 seconds.
+	DialTimeout time.Duration
+
+	// Logger used by the broker. By default all messages are dropped.
+	Log Logger
+}
+
+func NewBrokerConf(clientID string) BrokerConf {
+	return BrokerConf{
+		ClientID:         clientID,
+		DialTimeout:      time.Second * 10,
+		LeaderRetryLimit: 10,
+		LeaderRetryWait:  time.Millisecond * 500,
+		Log:              log.New(ioutil.Discard, "kafka", log.LstdFlags),
+	}
+}
+
+// Broker is abstract connection to kafka cluster, managing connections to all
+// kafka nodes.
+type Broker struct {
+	conf BrokerConf
+
+	mu       sync.Mutex
+	metadata clusterMetadata
+	conns    map[int32]*connection
+}
+
+// Dial connects to any node from given list of kafka addresses and after
+// successful metadata fetch, returns broker.
+// Returned broker is not initially connected to any kafka node.
+func Dial(nodeAddresses []string, conf BrokerConf) (*Broker, error) {
+	broker := &Broker{
+		conf:  conf,
+		conns: make(map[int32]*connection),
+	}
+
+	for _, addr := range nodeAddresses {
+		conn, err := NewConnection(addr, conf.DialTimeout)
+		if err != nil {
+			conf.Log.Printf("could not connect to %s: %s", addr, err)
+			continue
+		}
+		defer func(c *connection) {
+			_ = conn.Close()
```

Comment:
apologies, I was pointing to the wrong line before. It is this one that doesn't use the captured variable.